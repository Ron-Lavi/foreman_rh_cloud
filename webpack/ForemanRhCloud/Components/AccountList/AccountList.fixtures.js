import { noop } from 'patternfly-react';

export const API_SUCCESS_RESPONSE = {
  Account1: {
    label: 'test_org1',
    upload_report_status: 'running',
    generate_report_status: 'running',
  },
  Account2: {
    label: 'test_org2',
    upload_report_status: 'unknown',
    generate_report_status: 'failure',
  },
  Account3: {
    label: 'test_org3',
    upload_report_status: 'success',
    generate_report_status: 'running',
  },
};

export const accounts = API_SUCCESS_RESPONSE;

export const pollingProcessID = 0;

export const error = 'some-error';

export const accountID = 'user@redhat.com';

export const processStatusName = 'upload_report_status';

export const autoUploadEnabled = true;

export const props = {
  accounts,
  fetchAccountsStatus: noop,
  startAccountStatusPolling: noop,
  stopAccountStatusPolling: noop,
  pollingProcessID,
};
