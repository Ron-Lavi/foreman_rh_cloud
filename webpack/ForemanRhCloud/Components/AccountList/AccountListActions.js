import API from 'foremanReact/API';
import {
  INVENTORY_ACCOUNT_STATUS_POLLING,
  INVENTORY_ACCOUNT_STATUS_POLLING_ERROR,
  INVENTORY_ACCOUNT_STATUS_POLLING_START,
  INVENTORY_ACCOUNT_STATUS_POLLING_STOP,
  INVENTORY_PROCESS_RESTART,
} from './AccountListConstants';

export const fetchAccountsStatus = () => async dispatch => {
  try {
    const {
      data: { accounts, autoUploadEnabled },
    } = await API.get('accounts');
    dispatch({
      type: INVENTORY_ACCOUNT_STATUS_POLLING,
      payload: {
        accounts,
        autoUploadEnabled,
      },
    });
  } catch (error) {
    dispatch({
      type: INVENTORY_ACCOUNT_STATUS_POLLING_ERROR,
      payload: {
        error: error.message,
      },
    });
  }
};

export const startAccountStatusPolling = pollingProcessID => ({
  type: INVENTORY_ACCOUNT_STATUS_POLLING_START,
  payload: {
    pollingProcessID,
  },
});

export const stopAccountStatusPolling = pollingProcessID => dispatch => {
  clearInterval(pollingProcessID);
  dispatch({
    type: INVENTORY_ACCOUNT_STATUS_POLLING_STOP,
  });
};

export const restartProcess = (accountID, activeTab) => dispatch => {
  let processController = null;
  let processStatusName = null;

  if (activeTab === 'uploading') {
    processController = 'uploads';
    processStatusName = 'upload_report_status';
  } else {
    processController = 'reports';
    processStatusName = 'generate_report_status';
  }

  API.post(`${accountID}/${processController}`);
  dispatch({
    type: INVENTORY_PROCESS_RESTART,
    payload: {
      accountID,
      processStatusName,
    },
  });
};
