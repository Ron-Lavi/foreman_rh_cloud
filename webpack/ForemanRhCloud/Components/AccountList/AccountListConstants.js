export const INVENTORY_ACCOUNT_STATUS_POLLING =
  'INVENTORY_ACCOUNT_STATUS_POLLING';
export const INVENTORY_ACCOUNT_STATUS_POLLING_ERROR =
  'INVENTORY_ACCOUNT_STATUS_POLLING_ERROR';
export const INVENTORY_ACCOUNT_STATUS_POLLING_START =
  'INVENTORY_ACCOUNT_STATUS_POLLING_START';
export const INVENTORY_ACCOUNT_STATUS_POLLING_STOP =
  'INVENTORY_ACCOUNT_STATUS_POLLING_STOP';
export const INVENTORY_PROCESS_RESTART = 'INVENTORY_PROCESS_RESTART';
