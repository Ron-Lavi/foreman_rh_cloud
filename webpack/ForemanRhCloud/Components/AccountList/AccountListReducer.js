import Immutable from 'seamless-immutable';
import {
  INVENTORY_ACCOUNT_STATUS_POLLING,
  INVENTORY_ACCOUNT_STATUS_POLLING_ERROR,
  INVENTORY_ACCOUNT_STATUS_POLLING_START,
  INVENTORY_PROCESS_RESTART,
} from './AccountListConstants';
import { AUTO_UPLOAD_TOGGLE } from '../AutoUploadSwitcher/AutoUploadSwitcherConstants';

const initialState = Immutable({
  accounts: {},
  pollingProcessID: 0,
  error: null,
});

export default (state = initialState, action) => {
  const {
    payload: {
      pollingProcessID,
      error,
      accounts,
      accountID,
      processStatusName,
      autoUploadEnabled,
    } = {},
  } = action;

  switch (action.type) {
    case INVENTORY_ACCOUNT_STATUS_POLLING:
      return state.merge({
        ...state,
        accounts,
        autoUploadEnabled,
        error: null,
      });
    case INVENTORY_ACCOUNT_STATUS_POLLING_ERROR:
      return state.merge({
        ...state,
        accounts: {},
        error,
      });
    case INVENTORY_ACCOUNT_STATUS_POLLING_START:
      return state.merge({
        ...state,
        pollingProcessID,
      });
    case INVENTORY_PROCESS_RESTART:
      return state.setIn(['accounts'], {
        ...state.accounts,
        [accountID]: {
          ...state.accounts[accountID],
          [processStatusName]: 'Restarting...',
        },
      });
    case AUTO_UPLOAD_TOGGLE:
      return state.merge({
        ...state,
        autoUploadEnabled,
      });
    default:
      return state;
  }
};
