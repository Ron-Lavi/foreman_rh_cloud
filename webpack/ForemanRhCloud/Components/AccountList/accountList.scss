.account_list {
  .list-view-pf-expand {
    margin-top: 25px;
  }

  .list-group-item-container {
    min-height: 435px;
  }
}
