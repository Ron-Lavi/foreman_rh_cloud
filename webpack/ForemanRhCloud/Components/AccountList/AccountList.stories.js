import React from 'react';
import AccountList from './AccountList';

export default {
  title: 'AccountList',
  component: AccountList,
};

export const Basic = () => <AccountList />;
