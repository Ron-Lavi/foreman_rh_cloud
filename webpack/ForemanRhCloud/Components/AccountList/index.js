import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import * as actions from './AccountListActions';
import reducer from './AccountListReducer';
import AccountList from './AccountList';
import {
  selectAccounts,
  selectPollingProcessID,
  selectError,
} from './AccountListSelectors';

// map state to props
const mapStateToProps = state => ({
  accounts: selectAccounts(state),
  pollingProcessID: selectPollingProcessID(state),
  error: selectError(state),
});

// map action dispatchers to props
const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

// export reducers
export const reducers = { accountsList: reducer };

// export connected component
export default connect(mapStateToProps, mapDispatchToProps)(AccountList);
