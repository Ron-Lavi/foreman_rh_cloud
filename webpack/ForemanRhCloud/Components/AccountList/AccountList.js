import React, { Component } from 'react';
import { ListView, noop } from 'patternfly-react';
import PropTypes from 'prop-types';
import ListItem from './Components/ListItem';
import EmptyState from './Components/EmptyState';
import ErrorState from './Components/ErrorState';
import './accountList.scss';

class AccountList extends Component {
  componentDidMount() {
    const { fetchAccountsStatus, startAccountStatusPolling } = this.props;
    fetchAccountsStatus();
    const pollingProcessID = setInterval(fetchAccountsStatus, 5000);
    startAccountStatusPolling(pollingProcessID);
  }

  componentWillUnmount() {
    const { stopAccountStatusPolling, pollingProcessID } = this.props;
    stopAccountStatusPolling(pollingProcessID);
  }

  render() {
    const { accounts, error } = this.props;
    const accountIds = Object.keys(accounts);

    if (error) {
      return <ErrorState error={error} />;
    }

    if (accountIds.length === 0) {
      return <EmptyState />;
    }
    const items = accountIds.map((accountID, index) => {
      const account = accounts[accountID];
      return (
        <ListItem
          key={index}
          accountID={accountID}
          account={account}
          initExpanded={index === 0}
        />
      );
    });
    return <ListView className="account_list">{items}</ListView>;
  }
}

AccountList.propTypes = {
  fetchAccountsStatus: PropTypes.func,
  startAccountStatusPolling: PropTypes.func,
  stopAccountStatusPolling: PropTypes.func,
  pollingProcessID: PropTypes.number,
  account: PropTypes.shape({
    generate_report_status: PropTypes.string,
    upload_report_status: PropTypes.string,
  }),
  accounts: PropTypes.object,
  error: PropTypes.string,
};

AccountList.defaultProps = {
  fetchAccountsStatus: noop,
  startAccountStatusPolling: noop,
  stopAccountStatusPolling: noop,
  pollingProcessID: 0,
  account: {
    generate_report_status: 'unknown',
    upload_report_status: 'unknown',
  },
  accounts: {},
  error: '',
};

export default AccountList;
