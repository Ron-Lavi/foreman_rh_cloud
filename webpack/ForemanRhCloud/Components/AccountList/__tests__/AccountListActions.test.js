import { testActionSnapshotWithFixtures } from '@theforeman/test';
import {
  fetchAccountsStatus,
  startAccountStatusPolling,
  stopAccountStatusPolling,
  restartProcess,
} from '../AccountListActions';
import { pollingProcessID } from '../AccountList.fixtures';
import { accountID, activeTab } from '../../Dashboard/Dashboard.fixtures';

const fixtures = {
  'should fetchAccountsStatus': () => fetchAccountsStatus(),
  'should startAccountStatusPolling': () =>
    startAccountStatusPolling(pollingProcessID),
  'should stopAccountStatusPolling': () =>
    stopAccountStatusPolling(pollingProcessID),
  'should restartProcess': () => restartProcess(accountID, activeTab),
};

describe('AccountList actions', () => testActionSnapshotWithFixtures(fixtures));
