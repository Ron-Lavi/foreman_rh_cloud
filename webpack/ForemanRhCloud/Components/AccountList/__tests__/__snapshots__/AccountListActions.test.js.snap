// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountList actions should fetchAccountsStatus 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "error": "Cannot read property 'data' of undefined",
      },
      "type": "INVENTORY_ACCOUNT_STATUS_POLLING_ERROR",
    },
  ],
]
`;

exports[`AccountList actions should restartProcess 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "accountID": "some-account-ID",
        "processStatusName": "generate_report_status",
      },
      "type": "INVENTORY_PROCESS_RESTART",
    },
  ],
]
`;

exports[`AccountList actions should startAccountStatusPolling 1`] = `
Object {
  "payload": Object {
    "pollingProcessID": 0,
  },
  "type": "INVENTORY_ACCOUNT_STATUS_POLLING_START",
}
`;

exports[`AccountList actions should stopAccountStatusPolling 1`] = `
Array [
  Array [
    Object {
      "type": "INVENTORY_ACCOUNT_STATUS_POLLING_STOP",
    },
  ],
]
`;
